### PR TITLE
HSEARCH-2689 Rework how the retrieval of the EntityManagerFactory works

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-2689

So, this PR fixes multiple issues:

1. We used to rely on the contexts (job context, step context) way too much, to such a point that the jsr352-core project was actually jberet-specific. The JSR-352 spec is very conservative about how widely contexts are shared. Thus this PR takes the necessary steps to re-initialize the contexts whenever necessary.
2. We used to instantiate entity managers needlessly
3. We had some dead code (unused data in JobContextData)
3. We used to assign the singleton scope to some of our batch artifacts, which is utterly wrong and leads to inconsistent properties being used by these artifacts. Note that in order to fix this issue, we have to make the CDI integration JBeret-specific, because the Job and Partition CDI scopes are not defined in the JSR-352 spec, only in the JBeret implementation.

@mincong, if you have some time, could you please have a look and tell me something seems wrong to you? It can wait a few weeks, don't worry :)

### CODE REVIEW

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/EntityReader.java [hunk 1]
```diff
@@ -211,10 +242,30 @@ public void open(Serializable checkpointId) throws Exception {
...
+	 * data set up in the JobContextSetupListener (actually we can, but only with JBeret).
+	 * Thus we take care to re-create the data if necessary.
+	 *
+	 * See https://github.com/WASdev/standards.jsr352.jbatch/issues/50
+	 */
+	private JobContextData getJobContextData() throws ClassNotFoundException, IOException {
+		return JobContextUtil.getOrCreateData( jobContext,
+				emfRegistry, entityManagerFactoryScope, entityManagerFactoryReference,
+				entityTypes, serializedCustomQueryCriteria );
+	}
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/EntityReader.java:260]: I try to understand to concept. Here, we're aligned with the spec §9.4.1.1 about the lifecycle and the scope of `JobContext`, which thread-scoped. According to different implementations, the status of transient data model `JobContextData` varies:

- If we can re-use the job context data set up previously in the setup listener, then we do it
- If we cannot re-use the job context data set up previously in the setup listener, then we create a new one for this partition (thread).

Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/EntityReader.java:260]: Would it be a good idea to refactor the utility method `JobContextUtil#getOrCreateData` into a chaining method? However, the benefit of such refactoring is limited since it's only for internal usage, it's not heavily used, and this JSR-352 integration is not stable yet (so the refactor I suggest today might need to be changed again tomorrow).

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/PartitionContextData.java [hunk 2]
```diff
@@ -24,13 +20,6 @@
 
 	private PartitionProgress partitionProgress;
 
-	/**
-	 * Hibernate session, unwrapped from EntityManager. It is stored for sharing the session between item reader and
-	 * item processor. Notice that item reader and item processor of the same partition always run in the same thread,
-	 * so it should be OK. When the job stops, session object will be released before persisting this class's instance.
-	 */
-	private Session session;
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/PartitionContextData.java:32]: Great, we don't need to store the session here anymore :+1:

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/JobContextData.java [hunk 3]
```diff
@@ -60,6 +52,10 @@ public void setEntityManagerFactory(EntityManagerFactory entityManagerFactory) {
 		this.entityManagerFactory = entityManagerFactory;
 	}
 
+	public ExtendedSearchIntegrator getSearchIntegrator() {
+		return ContextHelper.getSearchIntegratorBySF( entityManagerFactory.unwrap( SessionFactory.class ) );
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/JobContextData.java:56]: :+1: :+1: :+1: Great! After the creation of this method, we don't need to instantiate entity managers needlessly.

The unwrapping method here makes me think about the usage the entity manager factory (EMF): we store the EMF in this class, and expose it via the getter method `getEntityManagerFactory()`. However, in the caller classes, EMF is systematically unwrapped into session factory (SF), so maybe we could store SF directly. What do you think?

I'm not asking for changes, because it's not relevant to the ticket. I evoked the discussion just to ensure we don't forget it.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import org.hibernate.search.jsr352.logging.impl.Log;
+import org.hibernate.search.jsr352.massindexing.impl.JobContextData;
+import org.hibernate.search.util.StringHelper;
+import org.hibernate.search.util.logging.impl.LoggerFactory;
+
+/**
+ * Utility allowing to set up and retrieve the job context data, shared by all the steps.
+ * <p>
+ * This is the default implementation, meant to be overridden (mainly its
+ * {@link #getEntityManagerFactoryRegistry()} method) when one needs to retrieve the entity manager
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java]: The method `#getEntityManagerFactoryRegistry()` does not exist. It was belong to the class `JobContextSetupListener`, but deleted in commit ab8afde, because the EMF registry can only be retrieved through injection now.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+			jobContext.setTransientUserData( data );
+		}
+		return data;
+	}
+
+	public static JobContextData getExistingData(JobContext jobContext) {
+		JobContextData data = (JobContextData) jobContext.getTransientUserData();
+		if ( data == null ) {
+			throw new AssertionFailure( "The job context data was unexpectedly missing;"
+					+ " there probably is something wrong with how Hibernate Search set up the job context data." );
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java:67]: Should this `AssertionFailure` be integrated into the JSR352 logger?

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/inject/scope/HibernateSearchJobScoped.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+/**
+ * Scope for job execution, to be mapped to the scopes specific to each JSR-352 implementation.
+ */
+@Target({TYPE})
+@Retention(RUNTIME)
+@Documented
+@Scope
+@Inherited
+public @interface HibernateSearchJobScoped {
```
Comment (mincong-h) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/inject/scope/HibernateSearchJobScoped.java:27]: I never create an annotation, so the following suggestion might be completely wrong...

I'm wondering if we can combine the two Hibernate Search scope annotations into only one, called `HibernateSearchScope`. Then add an enum attribute `Level level` inside, where the choices of this enum can be { JOB, PARTITION }? So we can use this annotation like:

```java
@Named(value = "org.hibernate.search.jsr352.massindexing.impl.JobContextSetupListener")
@HibernateSearchScope(level = Level.JOB)
public class JobContextSetupListener extends AbstractJobListener { ... }
```

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+ * </ul>
+ * Thus we use explicit type and scope registration as a workaround.
+ *
+ * @author Yoann Rodiere
+ */
+public class HibernateSearchJsr352Extension implements Extension {
+
+	public void afterTypeDiscovery(@Observes AfterTypeDiscovery event, BeanManager beanManager) {
+		registerType( event, beanManager, JobContextSetupListener.class );
+		registerType( event, beanManager, EntityReader.class );
```
Comment (mincong-h) [inline jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:60]: I'm trying to understand your code. Here, we register manually two class types, JobContextSetupListener and EntityReader, after the CDI component scanning. Or, in other words, after the type discovery in mode "annotated".

Such operation is necessary because we want to inject value into attributes of these classes. Therefore, they need to be instantiated before injection. For example, we want to inject the `EntityManagerFactoryRegistry` into the job setup listener, so the listener must be instantiated first. Normally, during CDI component scanning, the annotated bean can be discovered, register (instantiate) correctly. However, it isn't the case in WildFly (see [WFLY-8656](https://issues.jboss.org/browse/WFLY-8656)). So we have to create this extension class to handle the stuff manually.

Am I right?

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		event.addContext( new AliasedContext( alias, beanManager, target ) );
+	}
+
+	private static class AliasedContext implements Context {
+
+		private final Class<? extends Annotation> scopeType;
+		private final BeanManager targetBeanManager;
+		private final Class<? extends Annotation> targetScopeType;
+
+		public AliasedContext(Class<? extends Annotation> scopeType,
```
Comment (mincong-h) [inline jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:84]: The class is designed for internal usage, so maybe we could restrict the constructor visibility to private?

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+public class HibernateSearchJsr352Extension implements Extension {
+
+	public void afterTypeDiscovery(@Observes AfterTypeDiscovery event, BeanManager beanManager) {
+		registerType( event, beanManager, JobContextSetupListener.class );
+		registerType( event, beanManager, EntityReader.class );
+	}
+
+	public void afterBeanDiscovery(@Observes AfterBeanDiscovery event, BeanManager beanManager) {
+		addScopeAlias( event, beanManager, HibernateSearchJobScoped.class, JobScoped.class );
+		addScopeAlias( event, beanManager, HibernateSearchPartitionScoped.class, PartitionScoped.class );
```
Comment (mincong-h) [inline jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:65]: I'm trying to understand your code. Here, we add manually a scope alias after the bean discovery. Or, in other words, after starting a batch job execution. We manually associate the JBeret scopes to Hibernate Search scopes for our batch components:

- JBeret's `JobScoped` matches `HibernateSearchJobScoped`
- JBeret's `PartitionScoped` matches `HibernateSearchPartitionScoped`

We do this because the scopes of JBeret is an implementation-specific concept. As for other JSR-352 implementations, we need to have other alias mappings based on their concepts. And that's also the reason why you've renamed the module from `*-cdi` to `*-jberet`, because this CDI hack is only adapted to JBeret.

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+ * <li>In Wildfly, component scanning in dependencies is always done with the "all" discovery type,
+ * meaning that even components without CDI annotations will be discovered.
+ * Besides exposing private types unnecessarily,
+ * this is annoying because it puts default {@link EntityManagerFactoryRegistry} implementations into the CDI context,
+ * leading to conflicts when injecting such types
+ * (since this module, hibernate-search-jsr352-jberet, also provides an implementation).
+ * <br>See <a href="https://issues.jboss.org/browse/WFLY-8656">WFLY-8656</a>.
+ * <li>Even if we could solve the above, JBeret requires specific scope annotations to be set on the
+ * beans so that they are correctly scoped. Since the core isn't dependent on JBeret, we must specify these
+ * scopes manually when registering the beans.
```
Comment (mincong-h) [inline jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:50]: After having read your `<li>` above, I don't understand why my implementation worked anymore... Because I never add the JBeret scope annotations to ensure the ours beans are correctly scoped. Did it work because:

- The bean discovery mode was set to `all` previously?
- We didn't use `@Inject` to inject anything except the batch properties and JSR-352 contexts?

**Review by mincong-h — COMMENTED** [review]
Hi @yrodiere, I finished my review. Actually, it's more like a code learning than a review. Many thanks for this PR :100:

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/inject/scope/HibernateSearchJobScoped.java (hunk 6 above)
Comment (yrodiere) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/inject/scope/HibernateSearchJobScoped.java:27]: Conceptually that would be correct. However, technically that's not how scopes work. Each scope must have a single, distinct annotation to be able to use it as a CDI scope, and it's probably the same for Spring. The reason being that scoped are identified by the annotation *class*, and with your solution we'd have two scopes for the same annotation class.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/JobContextData.java (hunk 3 above)
Comment (yrodiere) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/JobContextData.java:56]: > maybe we could store SF directly. What do you think?

Right, we should make up your mind regarding which API we want to use. Created [HSEARCH-2706](https://hibernate.atlassian.net/browse/HSEARCH-2706).

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/EntityReader.java (hunk 1 above)
Comment (yrodiere) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/steps/lucene/EntityReader.java:260]: > I try to understand to concept. Here, [...]

Yep, that's correct.

> Would it be a good idea to refactor the utility method JobContextUtil#getOrCreateData into a chaining method

You mean, make it a builder? The thing is, there is no "optional" parameter here, so if we did use a builder, it would have to be a specially crafted one, with each method returning a different interface in order to ensure each parameter is filled in before we call `build()`. Which quite frankly is a bit much for a small utility like this one.

I agree the method is far from elegant, but as you said we might have to change it again soon (I'm thinking about [HSEARCH-2632](https://hibernate.atlassian.net/browse/HSEARCH-2632) in particular). So I'd say let's keep it that way and change it if possible in HSEARCH-2632.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java (hunk 5 above)
Comment (yrodiere) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java:67]: We could, but generally we don't care in cases like that. Thing is, this exception should really never occur except if *we* made an error in our code. That's why we throw an `AssertionFailure` instead of a `SearchException`, and that's why we don't consider internationalization (provided by the logger) a priority.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java (hunk 4 above)
Comment (yrodiere) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java]: Right, I'll fix that. Thanks!

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java (hunk 10 above)
Comment (yrodiere) [reply jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:50]: > We didn't use @Inject to inject anything except the batch properties and JSR-352 contexts?

That's right. Then I came and tried to integrate with CDI, and did it wrong with `@Singleton` scopes. It still worked, but only by chance, mainly because we only ever executed one job in each WildFly integration test.

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java (hunk 7 above)
Comment (yrodiere) [reply jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:60]: > Such operation is necessary because we want to inject value into attributes of these classes. Therefore, they need to be instantiated before injection. For example, we want to inject the EntityManagerFactoryRegistry into the job setup listener, so the listener must be instantiated first.

Right. CDI only injects beans that it created itself. When JBeret doesn't find a bean in the CDI context, it instantiates the bean itself, and thus the bean is not injected. So we really need the bean to be known from CDI and instantiated by CDI.

> Normally, during CDI component scanning, the annotated bean can be discovered, register (instantiate) correctly. However, it isn't the case in WildFly (see WFLY-8656). So we have to create this extension class to handle the stuff manually.

Right. To be precise, it *is* the case in WildFly (the beans are discovered), but WildFly is much too aggressive: it will discover not only the annotated beans, but *any* bean, even those we don't want to be in the CDI context. So we work it around by disabling automatic discovery and by registering the types manually.

> Am I right?

Yes.

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java (hunk 9 above)
Comment (yrodiere) [reply jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:65]: > Here, we add manually a scope alias after the bean discovery. Or, in other words, after starting a batch job execution

Actually this code is executed when CDI starts, which is only once and happens before a batch job is even started. But the alias will be effective for any job. We don't manage the scope lifecycle here, we just put some scope aliases; the actual scope lifecycle implementation is part of JBeret.

Apart from that, what you're correct.

File: jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java (hunk 8 above)
Comment (yrodiere) [reply jsr352/jberet/src/main/java/org/hibernate/search/jsr352/jberet/impl/HibernateSearchJsr352Extension.java:84]: It won't change anything: the class itself being private, the constructor won't be available outside of the surrounding class anyway. I feel making the constructor public makes it clearer that the surrounding class is expected to use it, though I'll grant you it won't change anything in practice (since the surrounding class is allowed to access private elements in the nested class, which is absurd if you ask me).

To sum up, it's a matter of taste...

**Review by yrodiere — COMMENTED** [review]
Thanks! I fixed a few issues you mentioned and merged.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java (hunk 5 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/massindexing/impl/util/JobContextUtil.java:67]: Understood, thanks!

**Comment by mincong-h** [comment]
Great work @yrodiere!! 🥇 🥇 🥇
